code/obj/item/gun/reagent.dm

```diff
@@ -0,0 +1,146 @@
+	var/capacity = 100 // reagent capacity of the gun
+	var/list/ammo_reagents = null // list of reagents accepted as ammo, leave blank if you want any to be accepted
+	var/projectile_reagents = 0 // whether the reagents should get transfered to the projectiles
+	var/dump_reagents_on_turf = 0 //set this to 1 if you want the dumped reagents to be put onto the turf instead of just evaporated into nothingness
+	var/custom_reject_message = "" //set this to a string if you want a custom message to be shown instead of the default when a reagent isnt accepted by the gun
+
+	New()
+		src.create_reagents(capacity)
+		..()
+
+	is_open_container()
+		return 1
+
+	alter_projectile(var/obj/projectile/P)
+		if(src.projectile_reagents && P && P.proj_data)
+			if (!P.reagents)
+				P.reagents = new /datum/reagents(P.proj_data.cost)
+				P.reagents.my_atom = P
+			src.reagents.trans_to(P, P.proj_data.cost)
+
+	on_reagent_change(add)
+		if(!add || !src.ammo_reagents)
+			src.update_icon()
+			return
+		var/mob/M = ismob(src.loc) ? src.loc : null
+		global.check_whitelist(src, src.ammo_reagents, M, src.custom_reject_message)
+		src.update_icon()
+
+	get_desc()
+		. = "[src.projectiles ? "It is set to [src.current_projectile.sname]. " : ""]There are [src.reagents.total_volume]/[src.reagents.maximum_volume] units left!"
+		if(src.current_projectile)
+			. += " Each shot will currently use [src.current_projectile.cost] units!"
+		else
+			. += "<span class='alert'>*ERROR* No output selected!</span>"
+		..()
+
+	update_icon()
+		return 0
+
+	canshoot()
+		if(src.reagents && src.current_projectile)
+			if(src.reagents.total_volume >= src.current_projectile.cost)
+				return 1
+		return 0
+
+	process_ammo(var/mob/user)
+		if (!canshoot())
+			boutput(user, "<span class='alert'>\The [src]'s internal reservoir does not contain enough reagents to fire it!</span>")
+		if(!src.projectile_reagents)
+			src.reagents.remove_any(src.current_projectile.cost)
+			src.update_icon()
+		return 1
+
+	MouseDrop(over_object, src_location, over_location)
+		..()
+		if(!isliving(usr))
+			return
+
+		if(get_dist(src, usr) > 1)
+			boutput(usr, "<span class='alert'>You need to be closer to empty \the [src] out!</span>")
+			return
+
+		if (!src.reagents)
+			boutput(usr, "<span class='alert'>The little cap on the fluid container is stuck. Uh oh.</span>")
+			return
+
+		if(src.reagents.total_volume)
+			if (src.dump_reagents_on_turf)
+				logTheThing("combat", usr, null, "transfers chemicals from [src] [log_reagents(src)] to [get_turf(src)] at [log_loc(usr)].")
+				src.reagents.trans_to(get_turf(src), src.reagents.total_volume)
+			src.reagents.clear_reagents()
+			src.update_icon()
+			boutput(usr, "<span class='notice'>You dump out \the [src]'s stored reagents.</span>")
+		else
+			boutput(usr, "<span class='alert'>There's nothing loaded to drain!</span>")
+
+	attackby(obj/item/I as obj, mob/user as mob)
+		if (istype(I, /obj/item/reagent_containers/glass))
+			return
```

Comment:
Why return?

Reply:
im not 100% sure, i copied what already worked from syringe guns / palis older code, but i think its because we want everything thats not a glass container to do what itd normally do, so everything thsts not glass makes the gun call back to the parent normally

but if its a glass container, we want to cut it short so it just transfers into it?? just an educated guess tho cos i didnt actually come up w the code

Reply:
That makes sense!